text/0161-telemetry-schema-and-resource.md

```diff
@@ -0,0 +1,196 @@
+renamed `cloud.zone` to `cloud.availability_zone`.  In a Schema world, this
+change could have been defined with the following schema rule:
+
+```yaml
+  next.version:
+    resources:
+      changes:
+        - rename_attributes:
+            attribute_map:
+              cloud.zone: cloud.availability_zone
+  previous.version:
+```
+
+If an exporter is defined expecting to find `cloud.zone` on `previous_version`,
+then when the SDK is bumped to `next.version` it will no longer be able to see
+`cloud.zone`.
+
+Specifically, we assume the following components could have independent
+versioning + release cycles:
+
+```
++-------------------+     +-----+     +----------+
+| Resource Detector | <-> | SDK | <-> | Exporter |
++-------------------+     +-----+     +----------+
+```
+
+## Explanation
+
+We propose a new set of SDK features:
+
+- A mechanism to "migrate" a resource up/down compatible schema changes.
+- A mechanism to determine if two schema URLs allow compatible changes.
+
+### Consuming resources at specific schema version
+
+The first is done by providing a new method against `Resource` in the SDK that
+will transform the attributes as necessary to match a given schema version.
+
+e.g. In Java you might have the following for resource:
+
+```java
+public abstract class Resource {
+    ... existing methods ...
+    /** The current schema version. */
+    public String getSchemaUrl();
+    /**
+     * Returns a new instance of Resource with labels compatible with the
+     * given schema url.
+     *
+     * <p> Returns an invalid Resource if conversion is not possible.
+     */
+    public Resource convertToSchema(String schemaUrl);
+}
+```
+
+This would allow a consumer of resources to ensure a stable view of resource
+labels, e.g.
+
+```java
+public static String MY_SCHEMA_VERSION="https://opentelemetry.io/schemas/1.1.0";
+public void myMetricMethod(MetricData metric) {
+    useStableResource(metric.getResource().convertToSchema(MY_SCHEMA_VERSION));
+}
+```
+
+### Understanding schema compatibility
+
+The second will be done via a new (public) SDK method that allows understanding
+whether or not two schemas are compatible, and which is the "newer" version.
+
+e.g. in Java you would have:
+
+```java
+public abstract class Schemas {
+    public static boolean isCompatible(String oldUrl, String newUrl);
+    public static boolean isNewer(String testUrl, String baseUrl);
+}
+```
+
+This could be used in the `Resource.merge` method to determine if resources
+can be converted to a compatible schema.  Specfically resource can be merged
+IFF:
+
+- The schemas are compatible
+- Converting all `Resource` instances to the "newest" schema version returns
+  valid `Resource` objects.
+
+## Internal details
+
+There are three major changes to SDKs for this proposal.
+
+### Understanding schema compatibility
+
+Here we propose simply providing an implementation of
+[SemVer](https://semver.org/) for isCompatible, and using the
```

Comment:
I'm a little confused on this. If you had a schema rule like the one above:

```yaml
  next.version:
    resources:
      changes:
        - rename_attributes:
            attribute_map:
              cloud.zone: cloud.availability_zone
  previous.version:
```

since there is a transform between versions, would the two schemas be compatible and have the same major version?

Reply:
It's not clear to me what the answer to this is, which is why I opened: https://github.com/open-telemetry/oteps/pull/162

IF the answer to the quesiton is "yes they're compatible", then I think this OTEP resolves *some* major pain that kind of change would cause.